File: problems/1182/solution.cpp
class Solution {
public:
    vector<int> shortestDistanceColor(vector<int>& colors, vector<vector<int>>& queries) {
        int n = size(colors);
        vector<int> ans;
        vector<vector<int>> dists(3, vector<int>(n, -1));
        
        for (int i = 0; i < n; i++) {
            for (int color = 1; color <= 3; color++) {
                if (color == colors[i]) {
                    dists[color - 1][i] = 0;
                } else if (i) {
                    dists[color - 1][i] = dists[color - 1][i - 1] == -1 ? -1 : 1 + dists[color - 1][i - 1];
                }
            }
        }
        
        for (int i = n - 1; i >= 0; i--) {
            for (int color = 1; color <= 3; color++) {
                if (color == colors[i]) {
                    dists[color - 1][i] = 0;
                } else if (i + 1 < n) {
                    int next = dists[color - 1][i + 1] == -1 ? -1 : 1 + dists[color - 1][i + 1];
                    
                    if (next != -1 and dists[color - 1][i] != -1) {
                        dists[color - 1][i] = min(dists[color - 1][i], next);
                    } else {
                        dists[color - 1][i] = max(dists[color - 1][i], next);
                    }
                }
            }
        }
        
        for (const vector<int> q: queries) {
            int i = q[0], c = q[1];
            
            ans.push_back(dists[c - 1][i]);
        }
        
        return ans;
    }
};

File: problems/1182/solution.go
package main
func shortestDistanceColor(colors []int, queries [][]int) []int {
    n, ans, dists := len(colors), make([]int, len(queries)), make([][]int, 3)
    
    for i := range dists {
        dists[i] = make([]int, n)
        
        for j := range dists[i] {
            dists[i][j] = -1
        }
    }
    
    for i := 0; i < n; i++ {
        for color := 1; color <= 3; color++ {
            if color == colors[i] {
                dists[color - 1][i] = 0;
            } else if i > 0 {
                if dists[color - 1][i - 1] != -1 {
                    dists[color - 1][i] = 1 + dists[color - 1][i - 1]
                }
            }
        }
    }
    
    for i := n - 1; i >= 0; i-- {
        for color := 1; color <= 3; color++ {
            if color == colors[i] {
                dists[color - 1][i] = 0
            } else if i + 1 < n {
                next := -1
                
                if dists[color - 1][i + 1] != -1 {
                    next = 1 + dists[color - 1][i + 1]
                }
                
                if next != -1 && dists[color - 1][i] != -1 {
                    dists[color - 1][i] = min(dists[color - 1][i], next)
                } else {
                    dists[color - 1][i] = max(dists[color - 1][i], next)
                }
            }
        }
    }
    
    for idx, query := range queries {
        i, c := query[0], query[1]
        
        ans[idx] = dists[c - 1][i]
    }
    
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    
    return b
}